servers/src/main/java/tachyon/worker/block/allocator/RoundRobinAllocator.java

```diff
@@ -0,0 +1,120 @@
+ * AnyTier or AnyDirInTier).
+ */
+public class RoundRobinAllocator implements Allocator {
+  private BlockMetadataManagerView mManagerView;
+
+  // We need to remember the last dir index for every storage tier
+  private Map<StorageTierView, Integer> mTierViewDirViews = new HashMap<StorageTierView, Integer>();
+
+  public RoundRobinAllocator(BlockMetadataManagerView view) {
+    mManagerView = Preconditions.checkNotNull(view);
+    
+    for (StorageTierView tierView : mManagerView.getTierViews()) {
+      mTierViewDirViews.put(tierView, -1);
+    }
+  }
+
+  @Override
+  public TempBlockMeta allocateBlockWithView(long userId, long blockId, long blockSize,
+      BlockStoreLocation location, BlockMetadataManagerView view) throws IOException {
+    mManagerView = view;
+    return allocateBlock(userId, blockId, blockSize, location);
+  }
+
+  /**
+   * Should only be accessed by {@link allocateBlockWithView} inside class.
+   * Allocates a block from the given block store location. The location can be a specific location,
+   * or {@link BlockStoreLocation#anyTier()} or {@link BlockStoreLocation#anyDirInTier(int)}.
+   *
+   * @param userId the ID of user to apply for the block allocation
+   * @param blockId the ID of the block
+   * @param blockSize the size of block in bytes
+   * @param location the location in block store
+   * @return a temp block meta if success, null otherwise
+   * @throws IOException if block location is invalid
+   */
+  private TempBlockMeta allocateBlock(long userId, long blockId, long blockSize,
+      BlockStoreLocation location) throws IOException {
+    if (location.equals(BlockStoreLocation.anyTier())) {
+      int tierAlias = 0; //always starting from the first tier
+      for (int i = 0; i < mManagerView.getTierViews().size(); i ++) {
+        StorageTierView tierView = mManagerView.getTierViews().get(tierAlias);
+        int dirViewIndex = getNextDirViewInTierView(tierView, blockSize);
+        if (dirViewIndex >= 0) {
+          mTierViewDirViews.put(tierView, dirViewIndex); // update
+          return tierView.getDirView(dirViewIndex).createTempBlockMeta(userId, blockId, blockSize);
+        } else { // we didn't find one in this tier, go to next tier
+          tierAlias ++;
+        }
+      }
+    } else if (location.equals(BlockStoreLocation.anyDirInTier(location.tierAlias()))) {
+      StorageTierView tierView = mManagerView.getTierView(location.tierAlias());
+      int dirViewIndex = getNextDirViewInTierView(tierView, blockSize);
+      if (dirViewIndex >= 0) {
+        mTierViewDirViews.put(tierView, dirViewIndex); // update
+        return tierView.getDirView(dirViewIndex).createTempBlockMeta(userId, blockId, blockSize);
+      }
+    } else {
+      StorageTierView tierView = mManagerView.getTierView(location.tierAlias());
+      StorageDirView dirView = tierView.getDirView(location.dir());
+      if (dirView.getAvailableBytes() >= blockSize) {
```

Comment:
What do you mean? @_@

Yes, getNextDirViewInTierView does iterate through all dirs (starting from
the last dir).





On 11 July 2015 at 16:28, Qifan Pu <notifications@github.com> wrote:

> In
> servers/src/main/java/tachyon/worker/block/allocator/RoundRobinAllocator.java
> <https://github.com/amplab/tachyon/pull/1040#discussion_r34417235>:
>
> > +          return tierView.getDirView(dirViewIndex).createTempBlockMeta(userId, blockId, blockSize);
> > +        } else { // we didn't find one in this tier, go to next tier
> > +          tierAlias ++;
> > +        }
> > +      }
> > +    } else if (location.equals(BlockStoreLocation.anyDirInTier(location.tierAlias()))) {
> > +      StorageTierView tierView = mManagerView.getTierView(location.tierAlias());
> > +      int dirViewIndex = getNextDirViewInTierView(tierView, blockSize);
> > +      if (dirViewIndex >= 0) {
> > +        mTierViewDirViews.put(tierView, dirViewIndex); // update
> > +        return tierView.getDirView(dirViewIndex).createTempBlockMeta(userId, blockId, blockSize);
> > +      }
> > +    } else {
> > +      StorageTierView tierView = mManagerView.getTierView(location.tierAlias());
> > +      StorageDirView dirView = tierView.getDirView(location.dir());
> > +      if (dirView.getAvailableBytes() >= blockSize) {
>
> Is it ever going to be true? My understanding is that
> getNextDirViewInTierView should have iterated through all dirViews.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/1040/files#r34417235>.
>



-- 
Tianyin XU,
http://cseweb.ucsd.edu/~tixu/